README.md
# FieldPalette

FieldPalette is a contao widget similar to [MultiColumnWizard] (https://github.com/menatwork/MultiColumnWizard).
Unlike MultiColumnWizard, fields are stored flatly into 'tl_fieldpalette' table and synced with its parent field.

The fieldpalette configuration is based on Contao's [Data Container Arrays](https://docs.contao.org/books/api/dca/index.html).

![alt fieldpalette wizard](./docs/img/fieldpalette_wizard.jpg)
*FieldPalette Wizard - ListView*

![alt fieldpalette edit](./docs/img/fieldpalette_edit.jpg)
*FieldPalette Wizard - Edit item*

## Technical instructions

### Default Setup (`tl_fieldpalette` table)

This example shows the setup of an fieldpalette field within tl_news by using it within an subpalette. That example is available within the module [heimrichhannot/contao-plus] (https://packagist.org/packages/heimrichhannot/contao-news_plus).

```
#/system/modules/[MyModule]/dca/tl_news.php

$dc = &$GLOBALS['TL_DCA']['tl_news'];

/**
 * Selectors
 */
$dc['palettes']['__selector__'][] = 'addVenues';

/**
 * Subpalettes
 */
$dc['subpalettes']['addVenues'] = 'venues';

/**
 * Fields
 */
$arrFields = array
(
	// venue
	'addVenues'           => array
	(
		'label'     => &$GLOBALS['TL_LANG']['tl_news']['addVenues'],
		'exclude'   => true,
		'inputType' => 'checkbox',
		'eval'      => array('submitOnChange' => true),
		'sql'       => "char(1) NOT NULL default ''",
	),
	'venues'              => array
	(
		'label'        => &$GLOBALS['TL_LANG']['tl_news']['venues'],
		'inputType'    => 'fieldpalette',
		'foreignKey'   => 'tl_fieldpalette.id',
		'relation'     => array('type' => 'hasMany', 'load' => 'eager'),
		'sql'          => "blob NULL",
		'fieldpalette' => array
		(
			'config' => array(
				'hidePublished' => false
			),
			'list'     => array
			(
				'label' => array
				(
					'fields' => array('venueName', 'venueStreet', 'venuePostal', 'venueCity'),
					'format' => '%s <span style="color:#b3b3b3;padding-left:3px">[%s, %s %s]</span>',
				),
			),
			'palettes' => array
			(
				'default' => 'venueName,venueStreet,venuePostal,venueCity,venueCountry,venueSingleCoords,venuePhone,venueFax,venueEmail,venueWebsite,venueText',
			),
			'fields'   => array
			(
				'venueName'         => array
				(
					'label'     => &$GLOBALS['TL_LANG']['tl_news']['venueName'],
					'exclude'   => true,
					'search'    => true,
					'inputType' => 'text',
					'eval'      => array('maxlength' => 255, 'tl_class' => 'long'),
					'sql'       => "varchar(255) NOT NULL default ''",
				),
				'venueStreet'       => array
				(
					'label'     => &$GLOBALS['TL_LANG']['tl_news']['venueStreet'],
					'exclude'   => true,
					'search'    => true,
					'inputType' => 'text',
					'eval'      => array('maxlength' => 255, 'tl_class' => 'w50'),
					'sql'       => "varchar(255) NOT NULL default ''",
				),
				'venuePostal'       => array
				(
					'label'     => &$GLOBALS['TL_LANG']['tl_news']['venuePostal'],
					'exclude'   => true,
					'search'    => true,
					'inputType' => 'text',
					'eval'      => array('maxlength' => 32, 'tl_class' => 'w50'),
					'sql'       => "varchar(32) NOT NULL default ''",
				),
				'venueCity'         => array
				(
					'label'     => &$GLOBALS['TL_LANG']['tl_news']['venueCity'],
					'exclude'   => true,
					'filter'    => true,
					'search'    => true,
					'sorting'   => true,
					'inputType' => 'text',
					'eval'      => array('maxlength' => 255, 'tl_class' => 'w50'),
					'sql'       => "varchar(255) NOT NULL default ''",
				),
				'venueCountry'      => array
				(
					'label'     => &$GLOBALS['TL_LANG']['tl_news']['venueCountry'],
					'exclude'   => true,
					'filter'    => true,
					'sorting'   => true,
					'inputType' => 'select',
					'options'   => System::getCountries(),
					'eval'      => array('includeBlankOption' => true, 'chosen' => true, 'tl_class' => 'w50'),
					'sql'       => "varchar(2) NOT NULL default ''",
				),
				'venueSingleCoords' => array
				(
					'label'         => &$GLOBALS['TL_LANG']['tl_news']['venueSingleCoords'],
					'exclude'       => true,
					'search'        => true,
					'inputType'     => 'text',
					'eval'          => array('maxlength' => 64),
					'sql'           => "varchar(64) NOT NULL default ''",
					'save_callback' => array
					(
						array('tl_news_plus', 'generateVenueCoords'),
					),
				),
				'venueText'         => array
				(
					'label'     => &$GLOBALS['TL_LANG']['tl_news']['venueText'],
					'exclude'   => true,
					'search'    => true,
					'inputType' => 'textarea',
					'eval'      => array('rte' => 'tinyMCE', 'tl_class' => 'clr'),
					'sql'       => "text NULL",
				),
			),
		),
	),
);

$dc['fields'] = array_merge($dc['fields'], $arrFields);
```

### Custom table setup (e.g. `tl_member_address`)

In order to use Fieldpalette with your own table, create a Data Container Array that extends from `$GLOBALS['TL_DCA']['tl_fieldpalette']`, as the following example describes.

```
<?php
/* /dca/tl_member_address.php */

\Controller::loadLanguageFile('tl_fieldpalette');
\Controller::loadDataContainer('tl_fieldpalette');
\Controller::loadDataContainer('tl_member');

$GLOBALS['TL_DCA']['tl_member_address'] = $GLOBALS['TL_DCA']['tl_fieldpalette'];
$dca                                    = &$GLOBALS['TL_DCA']['tl_member_address'];

$fields = [
    'company'     => $GLOBALS['TL_DCA']['tl_member']['fields']['company'],
    'phone'       => $GLOBALS['TL_DCA']['tl_member']['fields']['phone'],
    'fax'         => $GLOBALS['TL_DCA']['tl_member']['fields']['fax'],
    'street'      => $GLOBALS['TL_DCA']['tl_member']['fields']['street'],
    'street2'     => $GLOBALS['TL_DCA']['tl_member']['fields']['street2'],
    'postal'      => $GLOBALS['TL_DCA']['tl_member']['fields']['postal'],
    'city'        => $GLOBALS['TL_DCA']['tl_member']['fields']['city'],
    'state'       => $GLOBALS['TL_DCA']['tl_member']['fields']['state'],
    'country'     => $GLOBALS['TL_DCA']['tl_member']['fields']['country'],
    'addressText' => $GLOBALS['TL_DCA']['tl_member']['fields']['addressText'],
];

$dca['fields'] = array_merge($dca['fields'], $fields);
```

Than add the following fieldpalette input to your parent table (e.g. `tl_member`).

```
/* /dca/tl_member.php */

$dca = &$GLOBALS['TL_DCA']['tl_member'];

/**
* Adjust palettes
*/
$dca['palettes']['default'] = str_replace('country', 'country,additionalAddresses', $dca['palettes']['default']);

/**
* Adjust fields
*/
$dca['fields']['additionalAddresses'] = [
    'label'        => &$GLOBALS['TL_LANG']['tl_member']['additionalAddresses'],
    'inputType'    => 'fieldpalette',
    'foreignKey'   => 'tl_member_address.id',
    'relation'     => ['type' => 'hasMany', 'load' => 'eager'],
    'sql'          => "blob NULL",
    'fieldpalette' => [
        'config'   => [
            'hidePublished' => false,
            'table'         => 'tl_member_address',
        ],
        'list'     => [
            'label' => [
                'fields' => ['city'],
                'format' => '%s',
            ],
        ],
        'palettes' => [
            'default' => '{contact_legend},phone,fax;{address_legend},company,street,street2,postal,city,state,country,addressText',
        ],
    ],
];

```

### Additional dca reference

All attributes from https://docs.contao.org/books/api/dca/reference.html supported, if they are implemented yet. Additional attributes will be listed here

#### Listing records

##### Sorting

|  Key  | Value  | Description
|  ---  | ---  | ---
viewMode | View mode (integer) | **0** Table (default) <br /> **1** List 

### Support recursive copying of fieldpalette records by copying their parent record

Simply add a ```oncopy_callback``` to the dca containing fields of type "fieldpalette":

```
$GLOBALS['TL_DCA']['tl_*'] = array
(
    'config'   => array
    (
        // ...
        'oncopy_callback' => array(
            array('HeimrichHannot\FieldPalette\FieldPalette', 'copyFieldPaletteRecords')
        ),
    )
)
```

#### Manipulate fieldpalette records about to be copied on the fly

Sometimes your fieldpalette records contain references to other fieldpalette records. When copying them, reference ids don't match the new (copied) ids anymore.
You can adjust that by using the copy_callback definable in your field's dca (the field of type "fieldpalette"):

```
'inputType' => 'fieldpalette',
'eval'       => array(
    'fieldpalette' => array(
        'copy_callback' => array(
            array('tl_selection_model', 'updateOptionValuesOnCopy')
        )
    ),
    // ...
)
    
```

Example for such a callback:

```
public static function updateOptionValuesOnCopy($objFieldpalette, $intPid, $intNewId, $strTable, $arrData)
{
    $objFilter = \HeimrichHannot\FieldPalette\FieldPaletteModel::findByPk($objFieldpalette->selectionModel_questionData_options_filter);

    if ($objFilter === null)
        return;

    $objFilterNew = \HeimrichHannot\FieldPalette\FieldPaletteModel::findBy(
        array('selectionModel_questionData_filters_title=?', 'pid=?'),
        array($objFilter->selectionModel_questionData_filters_title, $intNewId)
    );

    if ($objFilterNew !== null)
    {
        $objFieldpalette->selectionModel_questionData_options_filter = $objFilterNew->id;
    }
}
```

## Features

### Widgets

Name | Description
---- | -----------
fieldpalette | The FieldPaletteWizard renders the tl_fieldpalette items and provide crud functionality within its parent record (e.g. tl_news).

### Fields

tl_fieldpalette:

Name | Description
---- | -----------
id | autoincrement unique identifiere
pid | id of the parent entry
ptable | parent table name (e.g. tl_news)
pfield | parent field name (e.g. tl_news.venues)
sorting | the sorting value
published | the published state (1 = published) 
start | timestamp from where the element is published 
stop | timestamp until the element is published

### Form Callbacks

tl_fieldpalette:

Type | Description
---- | -----------
oncreate_callback | Get fieldpalette key from request, check if the parent table is active within Fieldpalette Registry and set the pfield to tl_fieldpalette item. 
onsubmit_callback | Update/Sync parent fieldpalette item value (for example tl_news.venues) when tl_fieldpalette entries were updated.
oncut_callback | Update/Sync parent fieldpalette item value (for example tl_news.venues) when tl_fieldpalette entries were sorted.
ondelete_callback | Update/Sync parent fieldpalette item value (for example tl_news.venues) when tl_fieldpalette entries were deleted.


### Hooks

Name | Arguments | Description
---- | --------- | -----------
loadDataContainer | $strTable | Register fields from parent datacontainer (like tl_news) to tl_fieldpalette and disable fieldpalette support from back end modules where no fieldpalette fields exists (see: initializeSystem Hook). 
initializeSystem | - | Enable tl_fieldpalette table within all back end modules.	
executePostActions | $strAction, \DataContainer $dc | Add refreshFieldPaletteField ajax action that return the updated FieldPaletteWizard content.

## Restrictions

* only supports DC_Table DataContainers


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.4.10] - 2018-07-12

### Fixed
- Do not die() in `HookListener::executePostActionsHook` if `$field['inputType']` is not `fieldpalette`, otherwise nested field `executePostActions` Hooks wont work anymore (e.g. multicolumneditor) 

## [1.4.9] - 2018-07-03

### Fixed
- correctly size modal window responsive 

## [1.4.8] - 2018-03-02

### Fixed
- toggle icon permission check for custom tables 

## [1.4.7] - 2018-02-15

### Fixed
- toggleVisibility

## [1.4.6] - 2018-02-29

### Fixed
- contao 4 backend style

## [1.4.5] - 2017-11-29

### Fixed
- user and user group field permisssion in contao 4

## [1.4.4] - 2017-11-13

### Fixed
- required to old haste_plus version

## [1.4.3] - 2017-11-13

### Fixed
- fieldpalette is not allowed within `edit` mode for current back end module exception

## [1.4.2] - 2017-11-08

### Fixed
- `Fieldpalette::adjustBackendModules` should not execute `Controller::loadDataContainer()` as there is no `BackendUser` available, that is required for default values of `author` fields inside `tl_news` etc. 

## [1.4.1] - 2017-11-08

### Fixed
- `FieldPaletteHooks::extractTableFields` when fieldpalette dca is not loaded yet

## [1.4.0] - 2017-11-01

### Added
- Support for custom tables (instead of `tl_fieldpalette`), for more details check the README.

## [1.3.19] - 2017-09-05

### Fixed
- php Documentation for `FieldpaletteModel` `@return` value

## [1.3.18] - 2017-09-05

### Added
- `FieldpaletteModel::findPublishedByIds` added `columns` and `values` arguments that will be merged with statement
- `FieldpaletteModel::findPublishedByPidAndTableAndField` added `columns` and `values` arguments that will be merged with statement
- `FieldpaletteModel::findPublishedByPidsAndTableAndField` added `columns` and `values` arguments that will be merged with statement
- `FieldpaletteModel::findByPidAndTableAndField` added `columns` and `values` arguments that will be merged with statement

## [1.3.17] - 2017-09-05

### Fix
- Database updates not showing up in Contao 3 Install tool

## [1.3.16] - 2017-08-18

### Added
- `FieldpaletteModel::findPublishedByPidsAndTableAndField`

## [1.3.15] - 2017-08-18

### Fixed
- contao 4 closeModal fixed

## [1.3.14] - 2017-08-18

### Fixed
- contao 4 fieldpalette javascript compatibility

## [1.3.13] - 2017-08-18

### Fixed
- contao 4 dca-extractor handling for `tl_fieldpalette`

## [1.3.12] - 2017-08-16

### Fixed
- contao 4 compatibility, removed 'main.php' from button link if contao version > 4

## [1.3.11] - 2017-08-15

### Fixed
- contao 4 compatibility, add extra field palette fields

## [1.3.10] - 2017-07-27

### Fixed
- provide all available fieldpalette fields within `tl_user_group` to provide proper permission handling

## [1.3.9] - 2017-07-18

### Fixed
- PHP7 compatibility

## [1.3.8] - 2017-06-27

### Fixed
- array check in FieldPalette.php

## [1.3.7] - 2017-06-26

### Fixed
- fixed deps

## [1.3.6] - 2017-06-12

### Changed
- switched to heimrichhannot/datatables

## [1.3.5] - 2017-06-12

### Fixed
- array check in FieldPalette.php
- fixed config.php paths for optional contao 4 support
- removed traditional arrays

## [1.3.4] - 2017-05-11

### Fixed
- empty table error, within DC_Table::create

## [1.3.3] - 2017-05-09

### Fixed
- php 7 support, array handling in FieldPalette.php

## [1.3.2] - 2017-05-08

### Fixed
- `Fieldpalette::extractFieldPaletteFields` do not proceed when no array

## [1.3.1] - 2017-04-25

### Fixed
- list items had the wrong data container

## [1.3.0] - 2017-04-25

### Fixed
- nested fieldpalette issues

## [1.2.11] - 2017-04-12
- created new tag

## [1.2.10] - 2017-04-06

### Changed
- added php7 support, fixed contao-core dependency

## [1.2.9] - 2017-02-21

### Added
- Fixed saveNclose when contao referrer session failure

## [1.2.8] - 2017-02-21

### Added
- Handle saveNclose within fieldpalette modal

## [1.2.7] - 2017-01-10

### Fixed
- Add mandatory label to header field

## [1.2.6] - 2016-12-12

### Fixed
- fixed class issue in js

## [1.2.5] - 2016-12-12

### Changed
- removed bundled js libraries and referenced them by composer

## [1.2.4] - 2016-12-08

### Fixed
- fixed label handling

## [1.2.3] - 2016-12-05

### Changed
- jquery support within backend is now provided by haste_plus

## [1.2.2] - 2016-11-29

### Fixed
- fixed array check

## [1.2.1] - 2016-11-11

### Fixed
- disable internal cache for tl_fieldpalette dca container, otherwise dynamically added fields will be removed from database or not added

## [1.2.0] - 2016-10-26

### Added
- datatables support added to FieldPaletteWizard, it is now the default viewMode, can be changed within fieldpalette listing config `viewMode` parameter.
- datatables brings pagination, sort and search to fieldpalette widget

### Fixed
- Nested recursion error fixed, reworked the whole fieldpalette DCA registration
- Nested fieldpalettes fixed


dca/tl_fieldpalette.php
<?php
/**
 * Contao Open Source CMS
 *
 * Copyright (c) 2016 Heimrich & Hannot GmbH
 *
 * @package fieldpalette
 * @author  Rico Kaltofen <r.kaltofen@heimrich-hannot.de>
 * @license http://www.gnu.org/licences/lgpl-3.0.html LGPL
 */

$GLOBALS['TL_DCA']['tl_fieldpalette'] = [
    'config'      => [
        'dataContainer'     => 'Table',
        'ptable'            => '',
        'dynamicPtable'     => true,
        'fieldpalette'      => true, // required to grant access for back end modules
        'enableVersioning'  => true,
        'notCopyable'       => true,
        'onload_callback'   => [
            'setDateAdded'              => ['HeimrichHannot\HastePlus\Utilities', 'setDateAdded', true],
            'setReferrerOnSaveAndClose' => ['tl_fieldpalette', 'setReferrerOnSaveAndClose'],
        ],
        'sql'               => [
            'keys' => [
                'id'                                  => 'primary',
                'pid,ptable,pfield,published,sorting' => 'index',
            ],
        ],
        'oncreate_callback' => [
            ['tl_fieldpalette', 'setTable'],
        ],
        'onsubmit_callback' => [
            ['tl_fieldpalette', 'updateParentFieldOnSubmit'],
        ],
        'oncut_callback'    => [
            ['tl_fieldpalette', 'updateParentFieldOnCut'],
        ],
        'ondelete_callback' => [
            ['tl_fieldpalette', 'updateParentFieldonDelete'],
        ],
    ],
    'list'        => [
        'label'      => [
            'fields' => ['pid', 'ptable', 'pfield'],
            'format' => '%s <span style="color:#b3b3b3;padding-left:3px">[%s:%s]</span>',
        ],
        'operations' => [
            'edit'   => [
                'label' => &$GLOBALS['TL_LANG']['tl_fieldpalette']['edit'],
                'href'  => 'act=edit',
                'icon'  => 'edit.gif',
            ],
            'delete' => [
                'label'      => &$GLOBALS['TL_LANG']['tl_fieldpalette']['delete'],
                'href'       => 'act=delete',
                'icon'       => 'delete.gif',
                'attributes' => 'onclick="if(!confirm(\'' . $GLOBALS['TL_LANG']['MSC']['deleteConfirm']
                    . '\'))return false;FieldPaletteBackend.deleteFieldPaletteEntry(this,%s);return false;"',
            ],
            'toggle' => [
                'label'           => &$GLOBALS['TL_LANG']['tl_fieldpalette']['toggle'],
                'icon'            => 'visible.gif',
                'attributes'      => 'onclick="Backend.getScrollOffset();return AjaxRequest.toggleVisibility(this,%s)"',
                'button_callback' => ['tl_fieldpalette', 'toggleIcon'],
            ],
            'show'   => [
                'label' => &$GLOBALS['TL_LANG']['tl_fieldpalette']['show'],
                'href'  => 'act=show',
                'icon'  => 'show.gif',
            ],
        ],
    ],
    'palettes'    => [
        '__selector__' => ['published'],
    ],
    'subpalettes' => [
        'published' => 'start,stop',
    ],
    'fields'      => [
        'id'        => [
            'sql' => "int(10) unsigned NOT NULL auto_increment",
        ],
        'pid'       => [
            'sql' => "int(10) unsigned NOT NULL default '0'",
        ],
        'ptable'    => [
            'sql' => "varchar(64) NOT NULL default ''",
        ],
        'pfield'    => [
            'sql' => "varchar(255) NOT NULL default ''",
        ],
        'sorting'   => [
            'sql' => "int(10) unsigned NOT NULL default '0'",
        ],
        'tstamp'    => [
            'sql' => "int(10) unsigned NOT NULL default '0'",
        ],
        'dateAdded' => [
            'label'   => &$GLOBALS['TL_LANG']['MSC']['dateAdded'],
            'sorting' => true,
            'flag'    => 6,
            'eval'    => ['rgxp' => 'datim', 'doNotCopy' => true],
            'sql'     => "int(10) unsigned NOT NULL default '0'",
        ],
        'published' => [
            'exclude'   => true,
            'label'     => &$GLOBALS['TL_LANG']['tl_fieldpalette']['published'],
            'inputType' => 'checkbox',
            'eval'      => ['submitOnChange' => true, 'doNotCopy' => true],
            'sql'       => "char(1) NOT NULL default ''",
        ],
        'start'     => [
            'exclude'   => true,
            'label'     => &$GLOBALS['TL_LANG']['tl_fieldpalette']['start'],
            'inputType' => 'text',
            'eval'      => ['rgxp' => 'datim', 'datepicker' => true, 'tl_class' => 'w50 wizard'],
            'sql'       => "varchar(10) NOT NULL default ''",
        ],
        'stop'      => [
            'exclude'   => true,
            'label'     => &$GLOBALS['TL_LANG']['tl_fieldpalette']['stop'],
            'inputType' => 'text',
            'eval'      => ['rgxp' => 'datim', 'datepicker' => true, 'tl_class' => 'w50 wizard'],
            'sql'       => "varchar(10) NOT NULL default ''",
        ],
    ],
];

class tl_fieldpalette extends Backend
{
    /**
     * Import the back end user object
     */
    public function __construct()
    {
        parent::__construct();
        $this->import('BackendUser', 'User');
    }

    /**
     * Check permissions to edit table tl_fieldpalette
     */
    public function checkPermission()
    {
        // TODO
    }

    public function setReferrerOnSaveAndClose(\DataContainer $dc)
    {
        if (!isset($_POST['saveNclose'])) {
            return;
        }

        if (preg_match('/^contao/', TL_SCRIPT)) {
            $key = \Input::get('popup') ? 'popupReferer' : 'referer';
        }

        if ($key !== null) {
            $session = \Session::getInstance()->getData();

            if (!is_array($session[$key]) || !is_array($session[$key][TL_REFERER_ID])) {
                $session[$key][TL_REFERER_ID]['last'] = '';
            }

            while (count($session[$key]) >= 25) {
                array_shift($session[$key]);
            }

            $ref = \Input::get('ref');

            if ($ref != '' && isset($session[$key][$ref])) {
                if (!isset($session[$key][TL_REFERER_ID])) {
                    $session[$key][TL_REFERER_ID] = [];
                }

                $session[$key][TL_REFERER_ID]         = array_merge($session[$key][TL_REFERER_ID], $session[$key][$ref]);
                $session[$key][TL_REFERER_ID]['last'] = $session[$key][$ref]['current'];
            } elseif (count($session[$key]) > 1) {
                $session[$key][TL_REFERER_ID] = end($session[$key]);
            }

            $strUrl = substr(\Environment::get('requestUri'), strlen(TL_PATH) + 1);

            $session[$key][TL_REFERER_ID]['current'] = $strUrl;
            $session[$key][TL_REFERER_ID]['last']    = $strUrl;

            \Session::getInstance()->setData($session);
        }
    }

    public function setTable($strTable, $insertID, $arrSet, DataContainer $dc)
    {
        \Controller::loadDataContainer($strTable);

        if (!$GLOBALS['TL_DCA'][$strTable]['config']['fieldpalette']) {
            return;
        }


        $strFieldPalette = \HeimrichHannot\FieldPalette\FieldPalette::getPaletteFromRequest();

        $objModel        = \HeimrichHannot\FieldPalette\FieldPaletteModel::setTable($strTable)->findByPk($insertID);

        // if are within nested fieldpalettes set parent item tstamp
        if ($arrSet['ptable'] == 'tl_fieldpalette') {
            $objParent = \HeimrichHannot\FieldPalette\FieldPaletteModel::findByPk($objModel->pid);

            if ($objParent !== null) {
                $objParent->tstamp = time();
                $objParent->save();
            }
        }

        // set fieldpalette field
        $objModel->pfield = $strFieldPalette;
        $objModel->save();
    }

    /**
     * Return the "toggle visibility" button
     *
     * @param array $row
     * @param string $href
     * @param string $label
     * @param string $title
     * @param string $icon
     * @param string $attributes
     *
     * @return string
     */
    public function toggleIcon($row, $href, $label, $title, $icon, $attributes)
    {
        if (strlen(Input::get('tid'))) {
            $this->toggleVisibility(Input::get('tid'), (Input::get('state') === '1'), (@func_get_arg(12) ?: null));
            $this->redirect($this->getReferer());
        }

        // Check permissions AFTER checking the tid, so hacking attempts are logged
        if (!$this->User->hasAccess(\Config::get('fielpalette_table') . '::published', 'alexf')) {
            return '';
        }

        $href = \Haste\Util\Url::addQueryString('tid=' . $row['id'], $href);
        $href = \Haste\Util\Url::addQueryString('state=' . ($row['published'] ? '' : 1), $href);

        if (!$row['published']) {
            $icon = 'invisible.gif';
        }


        return '<a href="' . $href . '" title="' . specialchars($title) . '"' . $attributes . '>' . Image::getHtml(
                $icon,
                $label,
                'data-state="' . ($row['published'] ? 1 : 0) . '"'
            ) . '</a> ';
    }


    /**
     * Disable/enable a user group
     *
     * @param integer $intId
     * @param boolean $blnVisible
     * @param DataContainer $dc
     */
    public function toggleVisibility($intId, $blnVisible, DataContainer $dc = null)
    {
        // Set the ID and action
        Input::setGet('id', $intId);
        Input::setGet('act', 'toggle');

        if ($dc) {
            $dc->id = $intId; // see #8043
        }

        $this->checkPermission();

        // Check the field access
        if (!$this->User->hasAccess($dc->table . '::published', 'alexf')) {
            $this->log('Not enough permissions to publish/unpublish fieldpalette item ID "' . $intId . '"', __METHOD__, TL_ERROR);
            $this->redirect('contao/main.php?act=error');
        }

        $objVersions = new Versions($dc->table, $intId);
        $objVersions->initialize();

        // Trigger the save_callback
        if (is_array($GLOBALS['TL_DCA'][$dc->table]['fields']['published']['save_callback'])) {
            foreach ($GLOBALS['TL_DCA'][$dc->table]['fields']['published']['save_callback'] as $callback) {
                if (is_array($callback)) {
                    $this->import($callback[0]);
                    $blnVisible = $this->{$callback[0]}->{$callback[1]}($blnVisible, ($dc ?: $this));
                } elseif (is_callable($callback)) {
                    $blnVisible = $callback($blnVisible, ($dc ?: $this));
                }
            }
        }

        // Update the database
        $this->Database->prepare(
            "UPDATE " . $dc->table . " SET tstamp=" . time() . ", published='" . ($blnVisible ? '1' : '') . "' WHERE id=?"
        )->execute($intId);

        $objVersions->create();
        $this->log(
            'A new version of record "' . $dc->table . '.id=' . $intId . '" has been created' . $this->getParentEntries(
                $dc->table,
                $intId
            ),
            __METHOD__,
            TL_GENERAL
        );
    }

    public function updateParentFieldOnSubmit(DataContainer $dc)
    {
        $objCurrentRecord = \HeimrichHannot\FieldPalette\FieldPaletteModel::findByPk($dc->id);

        if ($objCurrentRecord === null) {
            return false;
        }

        $this->updateParentField($objCurrentRecord);
    }

    public function updateParentFieldOnCut(DataContainer $dc)
    {
        $objCurrentRecord = \HeimrichHannot\FieldPalette\FieldPaletteModel::findByPk($dc->id);

        if ($objCurrentRecord === null) {
            return false;
        }

        $this->updateParentField($objCurrentRecord);
    }

    public function updateParentFieldOnDelete(DataContainer $dc, $undoID)
    {
        $objCurrentRecord = \HeimrichHannot\FieldPalette\FieldPaletteModel::findByPk($dc->id);

        if ($objCurrentRecord === null) {
            return false;
        }

        $this->updateParentField($objCurrentRecord, $objCurrentRecord->id);
    }

    /**
     * Update the parent field with its tl_fieldpalette item ids
     *
     * @param DataContainer $dc
     *
     * @return bool
     */
    public function updateParentField($objCurrentRecord, $intDelete = 0)
    {
        $strClass = \Model::getClassFromTable($objCurrentRecord->ptable);

        if (!class_exists($strClass)) {
            return false;
        }

        /** @var \Model $strClass */
        $objParent = $strClass::findByPk($objCurrentRecord->pid);

        if ($objParent === null) {
            return false;
        }

        $objItems = \HeimrichHannot\FieldPalette\FieldPaletteModel::findByPidAndTableAndField(
            $objCurrentRecord->pid,
            $objCurrentRecord->ptable,
            $objCurrentRecord->pfield
        );

        $varValue = [];

        if ($objItems !== null) {
            $varValue = $objItems->fetchEach('id');

            // ondelete_callback support
            if ($intDelete > 0 && ($key = array_search($intDelete, $varValue)) !== false) {
                unset($varValue[$key]);
            }
        }

        if (empty($varValue)) {
            \Controller::loadDataContainer($objCurrentRecord->ptable);

            $arrData = $GLOBALS['TL_DCA'][$objCurrentRecord->ptable]['fields'][$objCurrentRecord->pfield];

            if (isset($arrData['sql'])) {
                $varValue = \Widget::getEmptyValueByFieldType($arrData['sql']);
            }
        }

        $objParent->{$objCurrentRecord->pfield} = $varValue;
        $objParent->save();
    }

}
